---
title: 'Analysis and Figures | Nature Scientific Data Manuscript'
author: 'Ismail Gbadamosi, Sandra Binias, BartÅ‚omiej Gielniewski, Ramiro Magno, Isabel Duarte, Ali Jawaid'
date: '`r Sys.Date()`'
output: bookdown::html_vignette2
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = '#>',
  message = FALSE,
  warning = FALSE,
  out.width = '100%',
  fig.width = 12,
  fig.height = 6
)
```

## Setup

Loading of required packages and data pre-calculated in other packages. The
pre-calculated data encompasses:

- tidy gene expression data: `tdp43.nsc34.rnaseq::gene_expression()`;
- samples metadata: `tdp43.nsc34.rnaseq::samples`;


```{r load_data}

# Package with RNA-seq data generated by the authors
# devtools::install_github("patterninstitute/tdp43")
library(tdp43.nsc34.rnaseq)

# Other packages
library(here)
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(patchwork)

## Create working objects

## Function to create a single tibble with expression in counts and TPMs, 
## and also with basic gene functional annotation from mgi
get_tpm_and_counts <- function() {
  ## Expression data normalized (TPMs)
  tpm <-
    tdp43.nsc34.rnaseq::gene_expression(unit = c("tpm"), format = "long") |>
    dplyr::left_join(tdp43.nsc34.rnaseq::samples[c("sample_id", "intervention")], by = "sample_id")
  
  ## Expression raw counts
  counts <-
    tdp43.nsc34.rnaseq::gene_expression(unit = "counts", format = "long") |>
    dplyr::left_join(tdp43.nsc34.rnaseq::samples[c("sample_id", "intervention")], by = "sample_id")
  
  ## Join counts and TPMs
  counts_tpm <- dplyr::full_join(counts,
                                 tpm,
                                 by = c("gene_name", "sample_id", "intervention"))
  
  ## Add gene biotype (coding or non-coding)
  counts_tpm2 <- dplyr::left_join(
    counts_tpm,
    tdp43.nsc34.rnaseq::mgi_annotation,
    by = dplyr::join_by(gene_name == marker_symbol),
    relationship = "many-to-many"
  ) |>
    dplyr::mutate(
      prot_coding = dplyr::case_when(
        biotype == "protein_coding" ~ "coding",
        is.na(biotype) ~ NA_character_,
        .default = "non_coding"
      )
    ) |>
    dplyr::relocate(intervention,
                    sample_id,
                    gene_name,
                    ensembl_gen_id,
                    biotype,
                    counts,
                    tpm)
  
  ## Return the merged tibble
  return(counts_tpm2)
}

# Get TPMs, Counts, and Annotation in a single table
exprs_counts_tpm <- get_tpm_and_counts()

## Counts matrix 
counts_matrix <- tdp43.nsc34.rnaseq::gene_expression(unit = c("counts"), format = "wide")

# Apply a regularized log transformation (a variance stabilizing transformation) to the gene expression count matrix (used for PCA, clustering analyses below). Please see ?DESeq2::rlog for more details.
 # NOTE: This step might take up to 1 minute. 
counts_rlg <- DESeq2::rlog(counts_matrix)

```


### Tardbp gene expression


Here we take a quick look at Tardbp expression (TPMs) across all samples,
colored by experimental condition. This allows the  validation of the experimental design, 
showing that the genetic interventions applied to the TDP43 gene in each experimental
group were successful.  


```{r exprs}

## Visualize Tardbp expression
exprs_counts_tpm |>
  filter(gene_name == 'Tardbp') |>
  ggplot(aes(x=sample_id, y=tpm)) +
  geom_col(aes(fill = intervention)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title='Tardbp expression')
```


### PCA Analysis

Confirm that the experimental conditions are the major sources of variation.

```{r pca}

# PCA calculation using the Regularized log transformed counts (variance stabilizing method))
pca <- prcomp(t(counts_rlg))

# Extract the components
components <- pca[["x"]]
components <- as_tibble(components, rownames = "sample_id")

# Add sample annotations to components
components_annot <- dplyr::left_join(components,
                                     as.data.frame(tdp43.nsc34.rnaseq::samples), 
                                     by = "sample_id") 

# Calculate the % variance per component
pca_percent_var <- round(pca$sdev^2/sum(pca$sdev^2)*100)


# 2D PCA | Using ggplo2

# Color by intervention
pca_intervention <-
  ggplot(components_annot, aes(x = PC1, y = PC2, color = intervention)) +
  geom_point(size = 3) +
  labs(
    title = "PCA | Colored by genetic intervention",
    x = paste0("PC1 (", pca_percent_var[1], "% variance)"),
    y = paste0("PC2 (", pca_percent_var[2], "% variance)")
  ) +
  theme_minimal()

# Print PCA plot
pca_intervention

```

### Sample Quality Control | Hierarchical clustering of the Correlation Matrix

```{r hcluster}
### Compute pairwise correlation values using the rlog counts matrix
rld_cor <- cor(counts_rlg)

### Plot heatmap using the correlation matrix
pheatmap::pheatmap(rld_cor)

```

### RNA-seq reads distribution and normalisation

Check that the normalization of the raw counts has been successful.

```{r distrib_and_normaliz}

# Boxplot
(exprs_counts_tpm |>
   filter(counts > 0) |>
  ggplot(aes(x = sample_id, y = log2(counts))) +
  geom_boxplot(aes(fill = intervention)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) 
) / 
 (exprs_counts_tpm |>
   filter(tpm > 0) |>
   ggplot(aes(x = sample_id, y = log2(tpm))) +
   geom_boxplot(aes(fill = intervention)) +
   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
 )


# Violin plot
(exprs_counts_tpm |>
    filter(counts > 0) |>
    ggplot(aes(x = sample_id, y = log2(counts))) +
    geom_violin(aes(fill = intervention)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) 
) / 
  (exprs_counts_tpm |>
     filter(tpm > 0) |>
     ggplot(aes(x = sample_id, y = log2(tpm))) +
     geom_violin(aes(fill = intervention)) +
     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
   )

# Density plot
(exprs_counts_tpm |>
     filter(counts > 0) |>
     ggplot(aes(x = log2(counts))) +
     geom_density(aes(fill = intervention), alpha = 0.5)
) / 
  (exprs_counts_tpm |>
     filter(tpm > 0) |>
     ggplot(aes(x = log2(tpm))) +
     geom_density(aes(fill = intervention), alpha=0.5)
  )


# Ridge plot
(exprs_counts_tpm |>
    filter(counts > 0) |>
    ggplot(aes(x = log2(counts), y = intervention, fill = intervention)) +
    ggridges::geom_density_ridges(alpha = 0.8) +
    # theme_minimal() +
    theme(axis.text.y = element_text(size = 12)) +
    theme(legend.position = "none")
) /
  (exprs_counts_tpm |>
     filter(tpm > 0) |>
     ggplot(aes(x = log2(tpm), y = intervention, fill = intervention)) +
     ggridges::geom_density_ridges(alpha=0.8) +
     # theme_minimal() +
     theme(axis.text.y = element_text(size = 12)) +
     theme(legend.position = "none")
  )

#
## Total number of genes measured
#
length(unique(exprs_counts_tpm$gene_name))

#
## How many genes have zero expression
#
exprs_counts_tpm |> group_by(gene_name) |> filter(counts == 0)

#
## How many genes have non-zero expression
#
exprs_counts_tpm |> group_by(gene_name) |> filter(counts > 0)

```

### Check gene's biotype

Verify that both coding and non-coding genes were present in the data.

```{r biotype}

# Count table
exprs_counts_tpm |>
  dplyr::distinct(gene_name, feature_type) |>
  dplyr::count(feature_type, sort = TRUE)


# Non_coding biotypes with small representativity (to change to other_non_coding)
non_cod_biotypes <- c("snoRNA gene", "unclassified non-coding RNA gene", 
                      "antisense lncRNA gene", "polymorphic pseudogene", 
                      "RNase P RNA gene", "snRNA gene", "telomerase RNA gene", 
                      "RNase MRP RNA gene", "DNA segment", "scRNA gene", 
                      "unclassified gene", "complex/cluster/region", "rRNA gene", 
                      "SRP RNA gene", "tRNA gene", "non-coding RNA gene")

# Plot
exprs_counts_tpm |>
  dplyr::distinct(gene_name, feature_type) |>
  na.omit() |>
  dplyr::mutate(feature_type = 
                  dplyr::case_when(feature_type %in% non_cod_biotypes ~ "other_non_coding",
                                   .default = feature_type)) |>
  dplyr::count(feature_type) |>
  ggplot(aes(x=reorder(feature_type, n), y=n)) +
  geom_col(aes(fill=feature_type)) +
  geom_text(aes(label = n), hjust = 0) +
  theme(legend.position="none") +
  xlab("Gene biotype") +
  ylab("Gene count") +
  coord_flip()
  
```



